src/lib/export/export.ts

```diff
@@ -0,0 +1,209 @@
+    options: ExportOptions,
+    outputOptions: any
+  ) {
+    this.filePath = filePath
+    this.connection = connection
+    this.table = table
+    this.filters = filters
+    this.options = options
+    this.outputOptions = outputOptions
+    this.id = this.generateId()
+  }
+
+  generateId(): string {
+    const md5sum = crypto.createHash('md5')
+
+    md5sum.update(Date.now().toString(), 'utf8')
+    md5sum.update(this.table.name)
+    md5sum.update(this.filePath)
+
+    return md5sum.digest('hex')
+  }
+
+  async getChunk(offset: number, limit: number): Promise<TableResult | undefined> {
+    const result = await this.connection.selectTop(
+      this.table.name,
+      offset,
+      limit,
+      [],
+      this.filters,
+      this.table.schema
+    );
+
+    return result
+  }
+
+  async getFirstRow() {
+    const row = await this.getChunk(0, 1)
+
+    if (row && row.result && row.result.length === 1) {
+      return row.result[0]
+    }
+  }
+
+  async writeToFile(content: string) {
+    return await fs.promises.appendFile(this.filePath, content + "\n")
+  }
+
+  async deleteFile() {
+    return await fs.promises.unlink(this.filePath)
+  }
+
+  async initExport(): Promise<void> {
+    this.status = Export.Status.Exporting
+    this.countExported = 0
+    
+    const firstRow = await this.getFirstRow()
+    const header = await this.getHeader(firstRow)
+    
+    await fs.promises.open(this.filePath, 'w+')
+
+    if (header) {
+      await this.writeToFile(header)
+    }
+  }
+
+  async exportData(): Promise<void> {
+      const chunk = await this.getChunk(this.countExported, this.options.chunkSize)
+
+      if (!chunk) {
+        this.status = Export.Status.Aborted
+        return
+      }
+
+      for (const formattedRow of this.formatChunk(chunk.result)) {
+        await this.writeToFile(formattedRow)
+      }
+
+      this.countTotal = chunk.totalRecords
+      this.countExported += chunk.result.length
+      const stats = await fs.promises.stat(this.filePath)
+      this.fileSize = stats.size
+      this.calculateTimeLeft()
+
+      if (this.countExported < this.countTotal && this.status === Export.Status.Exporting) {
+        await this.exportData()
+      }
+  }
+
+  async finalizeExport(): Promise<void> {
+    const footer = await this.getFooter()
+
+    if (footer) {
+      await this.writeToFile(footer)
+    }
+
+    this.status = Export.Status.Completed
+  }
+
+  async exportToFile(): Promise<void> {
+    try {
+      await this.initExport()
+      await this.exportData()
+
+      if (this.status === Export.Status.Aborted) {
+        if (this.options.deleteOnAbort) {
+          await this.deleteFile()
+        }
+        return
+      }
+
+      await this.finalizeExport()
+    } catch (error) {
+      this.status = Export.Status.Error
+      this.error = error
+
+      if (this.options.deleteOnAbort) {
+        await this.deleteFile()
+      }
+    }
+  }
+
+  calculateTimeLeft(): void {
```

Comment:
Ideally this would do the calculations based on the durations of the last n chunks instead.

Reply:
Totally fine I think except for MongoDB which famously slows down the deeper into an export you get :laughing:

Reply:
(we don't even support mongo yet)